File: services/geoapi/geoapi/HTTP/Handlers/RegisterAgentHandler.cs
using System.Net;
using System.Threading.Tasks;
using geoapi.Models;
using geoapi.Storages;
using geoapi.Utils;

namespace geoapi.HTTP.Handlers
{
    public class RegisterAgentHandler : BaseHandler
    {
        private readonly IStorage storage;
        private readonly ISettings settings;

        public RegisterAgentHandler(IStorage storage, ISettings settings) : base("POST", "agent")
        {
            this.storage = storage;
            this.settings = settings;
        }

        protected override async Task HandleRequestAsync(HttpListenerContext context)
        {
            var content = await context.Request.ReadContentAsync();

            var request = content.FromJson<RegisterAgentRequests>();

            var agent = new AgentInfo
            {
                AgentToken = GenerateId(settings.AgentIdSize),
                AgentName = request.AgentName,
            };

            storage.AddAgent(agent);

            await context.Response.Send(200, agent.ToJson());
            context.Response.Close();
        }

        private class RegisterAgentRequests
        {
            public string AgentName { get; set; }
        }

    }
}

File: services/geoapi/geoapi/HTTP/Handlers/RegisterAgentHandler.vb
Imports System
Imports System.Net
Imports System.Threading.Tasks
Imports geoapi.Models
Imports geoapi.Storages
Imports geoapi.Utils

Namespace HTTP.Handlers
    Public Class RegisterAgentHandler
        Inherits BaseHandler

        Private ReadOnly agentStorage As IAgentStorage
        Private ReadOnly settings As ISettings

        Public Sub New(ByVal agentStorage As IAgentStorage, ByVal settings As ISettings)
            MyBase.New("POST", "agent")
            Me.agentStorage = agentStorage
            Me.settings = settings
        End Sub

        Protected Overrides Async Function HandleRequestAsync(ByVal context As HttpListenerContext) As Task
            Dim content = Await context.Request.ReadContentAsync()
            Dim request = content.FromJson(Of RegisterAgentRequests)()
            Dim agent = New AgentInfo(GenerateId(settings.AgentIdSize), request.AgentName, DateTime.UtcNow + settings.TTL)
            agentStorage.AddAgent(agent)
            Await context.Response.Send(200, agent.ToJson())
            context.Response.Close()
        End Function
    End Class
End Namespace
